### ISSUE TITLE
Terraform apply crash (error) while creating private DNS a-records to existing private DNS zones.

### ISSUE THREAD

**Issue by Tafikay** [hashicorp/terraform-provider-azurerm#20592]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Community Note

* Please vote on this issue by adding a :thumbsup: [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform Version

Terraform v1.3.9

### AzureRM Provider Version

v3.44.1

### Affected Resource(s)/Data Source(s)

azurerm_private_dns_a_record

### Terraform Configuration Files

```hcl
I'm receiving issues during my Terraform apply.

I have 2 different repo's, 1 for my private dns zones and 1 for A-records.

In my HUB I have the following private-dns.tf module:
-dns.tf
resource "azurerm_private_dns_zone" "dns_zone" {
  count               = length(var.dns_zones)
  name                = var.dns_zones[count.index]
  resource_group_name = var.zone_rg
}


which creates my private DNS zones in my module call, example:
```
module "private-dns-zone-file" {
  source = "./modules/private-dns"
  location = "x"
  zone_rg = "x"
  dns_zones = ["privatelink.file.core.windows.net",]
}
```

In my landingzone:

```a-records.tf
  resource "azurerm_private_dns_a_record" "dns_a_record" {
    for_each = {
      for record in flatten([
        for dns in var.dns_a_records : [
          for a_record in dns.a_records : {
            dns_zone = dns.dns_zone
            name     = a_record.name
            ttl      = a_record.ttl
            ip       = a_record.ip_address
          }
        ]
      ]) : "${record.dns_zone}-${record.name}" => record
    }

    name                = each.value.name
    zone_name           = each.value.dns_zone
    resource_group_name = "x"
    ttl                 = each.value.ttl
    records             = [each.value.ip]
  }
```

my main.tf module call for my landing zone repo:

```
module "private_dns_a_records" {
	source          = "../../modules/a-records"

  providers = {
    azurerm.remote_sub = azurerm.remote_sub
  }

  dns_a_records = [
    {
      dns_zone     = "privatelink.azurewebsites.net",
      a_records       = [
        {
          name        = "test1",
          ip_address  = "10.0.0.5",
          ttl         = 5,
        },
        {
          name        = "test2",
          ip_address  = "10.0.0.6",
          ttl         = 5,
        },
      ]
    },
  ]
}

Note: the private DNS zones already exist and I receive an error during apply.
```

### Debug Output/Panic Output

```shell
│ Error: creating/updating Record Type (Subscription: "***"
│ Resource Group Name: "x"
│ Private Dns Zone Name: "privatelink.azurewebsites.net"
│ Record Type: "A"
│ Relative Record Set Name: "test1"): recordsets.RecordSetsClient#CreateOrUpdate: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceGroupNotFound" Message="Resource group 'x' could not be found."
│ 
│   with module.private_dns_a_records.azurerm_private_dns_a_record.dns_a_record["privatelink.azurewebsites.net-test1"],
│   on ../../modules/a-records/a-records.tf line 1, in resource "azurerm_private_dns_a_record" "dns_a_record":
│    1: resource "azurerm_private_dns_a_record" "dns_a_record" {
```

### Expected Behaviour

Create all my A-records in the list.

### Actual Behaviour

error:
│ Error: creating/updating Record Type (Subscription: "***"
│ Resource Group Name: "x"
│ Private Dns Zone Name: "privatelink.azurewebsites.net"
│ Record Type: "A"
│ Relative Record Set Name: "test1"): recordsets.RecordSetsClient#CreateOrUpdate: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceGroupNotFound" Message="Resource group 'x' could not be found."
│ 
│   with module.private_dns_a_records.azurerm_private_dns_a_record.dns_a_record["privatelink.azurewebsites.net-test1"],
│   on ../../modules/a-records/a-records.tf line 1, in resource "azurerm_private_dns_a_record" "dns_a_record":
│    1: resource "azurerm_private_dns_a_record" "dns_a_record" {

### Steps to Reproduce

Run my terraform code.

**Comment by Tafikay** [comment]
Note that both the resource group and private DNS zones do exist.

**Comment by aristosvo** [comment]
Is the resource group and private DNS zone in your "azurerm.remote_sub"? Confusion about the subscription might be the root cause of the error.

**Comment by Tafikay** [comment]
> 

Hey Aristosvo thank you for your reply, yes I'm 100% sure that the remote_sub is pointing towards the correct subscription. Also my plan is completely green showing me that everything is being set correctly (including RG)

**Comment by aristosvo** [comment]
Hi @Tafikay ! I understand that it is pointing to the right subscription, but is the correct alias provider also used to create the record? No `provider` is being set on the resource as far as I can see.

This is tested the easiest by adding in the data source for the resource group.

**Comment by Tafikay** [comment]
Aha, I have this set in main:
provider "azurerm" {
  features {}
  alias = "remote_sub"
  subscription_id = "xyz"
}

**Comment by CorrenSoft** [comment]
Try replacing `providers = {  azurerm.remote_sub = azurerm.remote_sub  }` with `providers = {  azurerm = azurerm.remote_sub  }` in your module call.
You are passing the provider defining an alias, but you are not declaring nor using that alias inside of your module, and Terraform instead of using that one is just grabbing the default.

**Comment by Tafikay** [comment]
@CorrenSoft good catch, that did the trick. very interesting. This gets me wondering now though, I first tried the same code from within my HUB Connectivity pipeline meaning that I did not need to use an azurerm provider yet I had the same behaviour, plan going green but failure during apply. Got any idea why that might have happened? Let me know if you need any more context and big thanks in advance.

**Comment by CorrenSoft** [comment]
Not sure... maybe the resource group or zone was not yet created? Without implicit or explicit dependency, terraform will try to create both and the same time... but I am just guessing.
I don't think that this is a version problem, unless the version is < 0.13.0 (I never used those so I don't know).

**Comment by manicminer** [comment]
Thanks @aristosvo and @CorrenSoft for assisting with troubleshooting. Since this seems to be a configuration issue rather than a bug with the provider, I'll close this one out for now.
